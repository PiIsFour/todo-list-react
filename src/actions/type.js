const actionType = Object.freeze({
	addList: 'ADD_LIST',
	selectList: 'SELECT_LIST',
	deleteList: 'DELETE_LIST',
})

export default actionType
