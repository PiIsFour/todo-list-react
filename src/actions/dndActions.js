import ActionType from './type'

export const DragItemType = Object.freeze({
	todoItem: 'TODO_ITEM',
})

export const reorderTodo = (item, to) => ({
	type: ActionType.reorderTodo,
	item,
	to,
})
