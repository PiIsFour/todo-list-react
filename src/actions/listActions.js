import { uuid } from 'uuidv4'
import actionType from './type'

export const addListWithId = (name, id) => ({
	type: actionType.addList,
	name,
	id,
})

export const addListFactory = uuid => name => addListWithId(name, uuid())

export const addList = addListFactory(uuid)

export const selectList = (id, name) => ({
	type: actionType.selectList,
	id,
	name,
})
