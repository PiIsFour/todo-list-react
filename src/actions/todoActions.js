import * as R from 'ramda'
import { uuid } from 'uuidv4'
import actionType from './type'

export const addTodoWithId = (name, toList, id) => ({
	type: actionType.addTodo,
	name,
	id,
	toList: R.pick(['id', 'name'], toList),
})

export const addTodoFactory = uuid => (name, toList) => addTodoWithId(name, toList, uuid())

export const addTodo = addTodoFactory(uuid)

export const checkTodo = (id, name, done) => ({
	type: actionType.checkTodo,
	id,
	name,
	done,
})
