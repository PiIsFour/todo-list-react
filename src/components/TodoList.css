.todo-list {
  --spacer: 2rem;

  grid-area: active;
  background: var(--clr-light);
	color: var(--clr-dark);
	position: relative;
}

.todo-list header {
  padding: var(--spacer);
  background: #e4e4e4;
  display: flex;
  align-items: center;
  justify-content: space-between;
}

.todo-list header h2 {
  margin: 0 1em 0 0;
}

.todo-list header p {
  margin: 0;
  font-size: 1rem;
}

.todo-list main {
  padding: var(--spacer);
}

.todo-list [type="checkbox"] {
  opacity: 0;
  position: absolute;
}

.todo-list .add-input button {
  color: var(--clr-primary);
}

.todo-list footer {
  display: flex;
  justify-content: space-evenly;
  position: absolute;
  width: 100%;
  left: 0;
  bottom: -35px;
  color: var(--clr-light);
}

.todo-list footer button {
	cursor: pointer;
  background: 0;
  border: 0;
  padding: 0;
  color: inherit;
  opacity: 0.7;
  font-size: 1rem;
  transition: color 200ms;
}

.todo-list footer button:hover {
  color: var(--clr-warning);
}
