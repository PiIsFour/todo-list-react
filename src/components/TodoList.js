import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import { useSelector, useDispatch } from 'react-redux'
import { useDrag, useDrop } from 'react-dnd'
import './TodoList.css'

import AddInput from './AddInput'
import { deleteList } from '../actions/listActions'
import { addTodo, checkTodo, clearCompletedTodos } from '../actions/todoActions'
import { DragItemType, reorderTodo } from '../actions/dndActions'

const TodoItem = ({id, name, done, dispatchCheckTodo, dispatchReorderTodo}) => {
	const ref = useRef()
	const [{isDragging}, drag] = useDrag({
		item: {
			type: DragItemType.todoItem,
			id,
			name,
		},
		collect: monitor => ({
			isDragging: !!monitor.isDragging(),
		}),
	})
	const [, drop] = useDrop({
		accept: DragItemType.todoItem,
		hover: (item) => isDragging ? null : dispatchReorderTodo(item, {id, name}),
	})
	drag(drop(ref))
	return <div className="todo" ref={ref}
		style={{
			opacity: isDragging ? 0.3 : 1,
		}}
	>
		<input
			type="checkbox"
			id={id}
			checked={done}
			onChange={e => dispatchCheckTodo(id, name, e.target.checked)}
		/>
		<label htmlFor={id}>
			<span className="custom-checkbox"></span>
			{name}
		</label>
	</div>
}

const renderTodo = ({dispatchCheckTodo, dispatchReorderTodo}) => ({id, name, done}) =>
	<TodoItem id={id} key={id} name={name} done={done}
		dispatchCheckTodo={dispatchCheckTodo}
		dispatchReorderTodo={dispatchReorderTodo}
	/>

export const TodoListView = ({selectedList, dispatch}) => {
	if(!selectedList) return null
	const { id, name, todos } = selectedList
	const remainingTodos = todos.filter(({done}) => !done).length
	const dispatchDeleteList = () => dispatch(deleteList(id, name))
	const dispatchAddTodo = (name) => dispatch(addTodo(name, selectedList))
	const dispatchCheckTodo = (id, name, done) => dispatch(checkTodo(id, name, done))
	const dispatchClearCompletedTodos = () => dispatch(clearCompletedTodos(selectedList))
	const dispatchReorderTodo = (item, to) => dispatch(reorderTodo(item, to))
	return <section className="todo-list">
		<header>
			<h2>{name}</h2>
			<p>{`${remainingTodos} ${remainingTodos===1 ? 'task' : 'tasks'} remaining`}</p>
		</header>
		<main>
			<div className="todos">
				{todos.map(renderTodo({dispatchCheckTodo, dispatchReorderTodo}))}
			</div>
			<AddInput lable="new task name" onAdd={dispatchAddTodo} />
		</main>
		<footer>
			<button onClick={dispatchClearCompletedTodos}>Clear completed tasks</button>
			<button onClick={dispatchDeleteList}>Delete list</button>
		</footer>
	</section>
}

TodoListView.propTypes = {
	selectedList: PropTypes.shape({
		id: PropTypes.string.isRequired,
		name: PropTypes.string.isRequired,
		todos: PropTypes.arrayOf(PropTypes.shape({
			id: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired,
			done: PropTypes.bool.isRequired,
		})).isRequired,
	}),
	dispatch: PropTypes.func,
}

export const getSelectedList = ({todoLists, selectedList}) => todoLists.find(({id}) => id === selectedList.id)

const TodoList = () => {
	const selectedList = useSelector(getSelectedList)
	const dispatch = useDispatch()
	return <TodoListView selectedList={selectedList} dispatch={dispatch} />
}

export default TodoList
