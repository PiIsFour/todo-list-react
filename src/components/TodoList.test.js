import React from 'react'
import { shallow } from 'enzyme'

import { deleteList } from '../actions/listActions'
import { addTodoWithId } from '../actions/todoActions'

import { TodoListView } from './TodoList'

describe('ListsOverview', () => {
	it('snapshot matches', () => {
		const selectedList = {
			id: '12345',
			name: 'test list name',
			todos: [],
		}
		const wrapper = shallow(<TodoListView selectedList={selectedList} />)
		expect(wrapper).toMatchSnapshot()
	})

	it('shows name of selected list', () => {
		const selectedList = {
			name: 'test list name',
			todos: [],
		}
		const wrapper = shallow(<TodoListView selectedList={selectedList} />)
		expect(wrapper.find('header h2')).toHaveText('test list name')
	})

	it('shows # of remaining todos if 0', () => {
		const selectedList = {
			todos: [],
		}
		const wrapper = shallow(<TodoListView selectedList={selectedList} />)
		expect(wrapper.find('header p')).toHaveText('0 tasks remaining')
	})

	it('shows # of remaining todos if 1', () => {
		const selectedList = {
			todos: [{done: true}, {done: true}, {done: false}],
		}
		const wrapper = shallow(<TodoListView selectedList={selectedList} />)
		expect(wrapper.find('header p')).toHaveText('1 task remaining')
	})

	it('dispatches deleteList action', () => {
		const selectedList = {
			id: '12345',
			name: 'test list name',
			todos: [],
		}
		const dispatch = jest.fn()
		const wrapper = shallow(<TodoListView selectedList={selectedList} dispatch={dispatch} />)
		const deleteButton = wrapper.find('footer').childAt(1)
		expect(deleteButton).toHaveText('Delete list')
		deleteButton.simulate('click')
		expect(dispatch).toHaveBeenCalledWith(deleteList('12345', 'test list name'))
	})

	it('dispatches a addTodo action', () => {
		const selectedList = {
			id: '12345',
			name: 'test list',
			todos: [],
		}
		const dispatch = jest.fn()
		const wrapper = shallow(<TodoListView selectedList={selectedList} dispatch={dispatch} />)
		wrapper.find('AddInput').prop('onAdd')('new todo name')
		expect(dispatch).toHaveBeenCalledWith(
			addTodoWithId('new todo name', {id: '12345', name: 'test list'}, expect.toBeAUuid()),
		)
	})

	it('getSelectedList returns the selected list', () => {
		const state = {
			todoLists: [
				{id: '12'},
				{id: '42'},
				{id: '152'},
			],
			selectedList: {id: '42'},
		}
		expect(getSelectedList(state)).toBe(state.todoLists[1])
	})
})
