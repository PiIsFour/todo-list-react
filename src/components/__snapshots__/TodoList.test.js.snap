// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListsOverview snapshot matches 1`] = `
<section
  className="todo-list"
>
  <header>
    <h2>
      test list name
    </h2>
    <p>
      1 task remaining
    </p>
  </header>
  <main>
    <div
      className="todos"
    >
      <TodoItem
        dispatchCheckTodo={[Function]}
        dispatchReorderTodo={[Function]}
        done={false}
        id="42"
        key="42"
        name="hello world"
      />
    </div>
    <AddInput
      lable="new task name"
      onAdd={[Function]}
    />
  </main>
  <footer>
    <button
      onClick={[Function]}
    >
      Clear completed tasks
    </button>
    <button
      onClick={[Function]}
    >
      Delete list
    </button>
  </footer>
</section>
`;
